### ISSUE TITLE
Inconsistent prediction by model using Attention layer

### ISSUE THREAD

**Issue by prachi92sharma** [tensorflow/tensorflow#40911]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS Catalina 10.15.5
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: NA
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.2.0
- Python version: 3.7.7
- Bazel version (if compiling from source): NA
- GCC/Compiler version (if compiling from source): NA
- CUDA/cuDNN version: NA
- GPU model and memory: NA

**Context**
While building a text classification model, we are using `tf.keras.layers.Attention`. Inputs are of the type 1d numpy array, each of the same shape padded by 0 at the end, and all starts with a start token, currently chosen as 1. The values are ID assigned to different words in the vocabulary.

**Describe the expected behavior**
Suppose for 4 sentences, inputs are a, b, c, d. Then we expect that `model.predict([a,b,c,d])` will be same as `np.concatenate([model.predict(np.expand_dims(a, axis=0), np.expand_dims(b, axis=0), np.expand_dims(c, axis=0), np.expand_dims(d, axis=0)], axis=0)`. 

**Describe the current behavior**
But we are not getting the expected result and there is a significant difference between the predicted probabilities. Sometimes even the position of the maximum probability will vary. So the result using `argmax` is not consistent.

**Observations**
1. If we remove Attention layer this inconsistency won't be there. 
2. Inconsistency is not affected by permutation. `[a, b, c, d]` will give same result as `[c, a, b, d]`.

**Standalone code to reproduce the issue**
```python
import numpy as np
from tensorflow import keras

tr_x = np.random.randint(0, 1024, size=(200, 50))
tr_y = np.random.randint(0, 4, 200)

il = keras.Input(shape=(50,))
el = keras.layers.Embedding(1024, 256)(il)
gl_out, gl_st = keras.layers.GRU(64, return_sequences=True, return_state=True)(el)
al = keras.layers.Attention()([gl_out, gl_st])
gl = keras.layers.GRU(16)(al)
ol = keras.layers.Dense(4, activation="softmax")(gl)

m = keras.Model(inputs=il, outputs=ol)
m.compile("adam", "sparse_categorical_crossentropy")

m.fit(tr_x, tr_y, epochs=16)

# Checking for inconsistency
ts_x = np.random.randint(0, 1024, size=(4, 50))
print(m.predict(ts_x))
print(m.predict(ts_x[:1]))
print(m.predict(ts_x[1:2]))
print(m.predict(ts_x[2:3]))
print(m.predict(ts_x[3:]))

# Checking for Permutation
x = np.random.randint(0, 1024, size=(8, 50))
p = np.random.permutation(8)
a = m.predict(x[p])
b = m.predict(x)[p]
np.testing.assert_allclose(a, b, 1e-3)
```

Is it an expected behavior or we are missing something in the code?

Thanks

**Comment by ravikyram** [comment]
@prachi92sharma 

I have tried in colab with TF 2.2,nightly versions.Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/889fb09fc6a01cd2a138c3c855ecb851/untitled68.ipynb).You also seeing the same behavior?.Thanks!

**Comment by prachi92sharma** [comment]
@ravikyram yes, I am also seeing this behavior. When passing ts_x(test data) in a single batch, predicted probabilities are different from the one when each data point in test data is passed individually.

**Comment by prachi92sharma** [comment]
@ravikyram @gowthamkpr @qlzh727 : Any update on this issue?

**Comment by yarnabrina** [comment]
@qlzh727, I'm also facing the same issue. Is there any solution/workaround for this?

**Comment by sushreebarsa** [comment]
Was able to replicate the issue with TF 2.5,please find the gist[ here](https://colab.research.google.com/gist/sushreebarsa/25cd92aa4db6677ded442bfe860134df/untitled322.ipynb) ..Thanks!

**Comment by kodayu** [comment]
Can anyone please tell me what the problem is? I have been following this problem for a long time but it has not been solved. Hope to get guidance!

**Comment by yarnabrina** [comment]
@kodayu My expectation is `f([x, y, z])` equals `[f(x), f(y), f(z)]`, and that's not the case. Here `f` is the prediction function, of course.

The reason for to this expectation is that prediction of different inputs are independent, so I don't understand why the difference occurs.

**Comment by tensorflowbutler** [comment]
Hi There,

 This is a stale issue. As you are using an older version of tensorflow, we are checking to see if you still need help on this issue. Please test the issue with the latest TensorFlow (TF2.7 and tf-nightly). If the issue still persists with the newer versions of TF, please feel free to open it in [keras-team/keras](https://github.com/keras-team/keras/issues) repository by providing details about the issue and a standalone code to reproduce the issue. Thanks! 

 Please note that Keras development has moved to a separate Keras-team/keras repository to focus entirely on only Keras. Thanks!
